data-raw/healthy-places/travel-time-to-gp.R

```diff
@@ -0,0 +1,233 @@
+library(tidyverse)
+library(geographr)
+library(osmdata)
+library(jsonlite)
+library(httr)
+library(sf)
+library(traveltimeR)
+library(readxl)
+
+Sys.setenv(TRAVELTIME_ID = "<INSERT YOUR ID HERE>")
+Sys.setenv(TRAVELTIME_KEY = "<INSERT YOUR API KEY HERE>")
+
+# ---- Fetch GPs in Northern Ireland ----
+# Bounding box for Northern Ireland
+ni_bb <-
+  getbb("Northern Ireland")
+
+# Search for GPs
+GPs <- opq(ni_bb, timeout = 1000) %>%
+  add_osm_feature("amenity", "doctors") %>%
+  osmdata_sf()
+
+# Get Local Authorities in Great Britain for the next step
+ni_lad <-
+  boundaries_ltla21 |>
+  filter(str_detect(ltla21_code, "^N"))
+
+# Some GPs are not in Northern Ireland - remove them from the dataset
+ni_GPs <-
+  GPs$osm_points[ni_lad, ] |>
+  select(osm_id)
+
+# - Test plot to check GP locations -
+# ni_lad |>
+#   ggplot() +
+#   geom_sf(
+#     fill = NA,
+#     colour = "black"
+#   ) +
+#   geom_sf(
+#     data = ni_GPs,
+#     inherit.aes = FALSE
+#   )
+
+# Calculate neighbours for each Local Authority
+neighbours <- st_touches(ni_lad)
+
+# Look up the Local Authority each GP is in
+ni_GPs_lad <-
+  st_join(ni_GPs, ni_lad) |>
+  filter(!is.na(ltla21_code)) |>
+  mutate(
+    # Round coords to 3 decimal points to save memory
+    lat = st_coordinates(geometry)[,2] |> round(3),
+    lng = st_coordinates(geometry)[,1] |> round(3)
+  ) |>
+  st_drop_geometry() |>
+  as_tibble() |>
+  select(
+    ltla21_code,
+    osm_id,
+    lat,
+    lng
+  )
+
+# ---- Population-weighted centroids for Super Data Zones ----
+# Source: https://www.nisra.gov.uk/publications/census-2021-population-weighted-centroids-data-zones-and-super-data-zones
+GET(
+  "https://www.nisra.gov.uk/system/files/statistics/geography-census-2021-population-weighted-centroids-for-data-zones-and-super-data-zones.xlsx",
+  write_disk(tf <- tempfile(fileext = ".xlsx"))
+)
+
+sdz21_centroids_raw <- read_excel(tf, sheet = "SDZ2021")
+
+sdz21_centroids <-
+  sdz21_centroids_raw |>
+  st_as_sf(coords = c("X", "Y"), crs = 29902) |>  # The coordinates are in Irish Grid
+  st_transform(crs = 4326) |>
+  mutate(
+    # Round coords to 3 decimal points to save memory
+    lat = st_coordinates(geometry)[,2] |> round(3),
+    lng = st_coordinates(geometry)[,1] |> round(3)
+  ) |>
+  st_drop_geometry() |>
+  select(
+    sdz21_code = SDZ2021_code,
+    lat,
+    lng
+  )
+
+# ---- Calculate travel time between data zones and GPs ----
```

Comment:
As I don't have traveltime credentials, I can't run this code to check the working. But, scanning the code, the logic looks sound.

I do however, recommend refactoring this code. It uses a for loop nested within a for loop, with `GP_travel_time` not being preallocated a size before the loop. As far as I am aware, for loops without preallocation can be slow due to dynamic memory reallocation. When you don't preallocate the output vector, R needs to repeatedly resize the vector, which involves copying the entire data structure each iteration. This results in slow code. Additionally, the logic is quite hard to follow (and will be hard to debug).

I recommend:
- splitting the logic into functions
- using `purrr::map()` and family in place of the for loops (or you can use base `Map()` or `lapply()`)

Reply:
That's a great idea. Let's explore this next year. The need to `Sys.sleep(2)` slows the loop down anyway, so there will be a limit to how fast it can run. I'm not a strong enough functional programmer to be able to convert the for loops while making sure that running `time_filter_fast(...)` via `purrr:map()` doesn't exceed our API limits.